sklego/meta/zero_inflated_regressor.py

```diff
@@ -0,0 +1,108 @@
+import numpy as np
+from sklearn.base import BaseEstimator, RegressorMixin, clone
+from sklearn.utils.validation import check_is_fitted, check_X_y, check_array
+
+
+class ZeroInflatedRegressor(BaseEstimator, RegressorMixin):
+    """
+    A meta regressor for zero-inflated datasets, i.e. the targets contain a lot of zeroes.
+
+    `ZeroInflatedRegressor` consists of a classifier and a regressor.
+
+        - The classifier's task is to find of if the target is zero or not.
+        - The regressor's task is to output a (usually positive) prediction whenever the classifier indicates that the there should be a non-zero prediction.
+
+    The regressor is only trained on examples where the target is non-zero, which makes it easier for it to focus.
+
+    At prediction time, the classifier is first asked if the output should be zero. If yes, output zero.
+    Otherwise, ask the regressor for its prediction and output it.
+
+    Parameters
+    ----------
+    classifier : Any, scikit-learn classifier
+        A classifier that answers the question "Should the output be zero?".
+
+    regressor : Any, scikit-learn regressor
+        A regressor for predicting the target. Its prediction is only used if `classifier` says that the output is non-zero.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.ensemble import ExtraTreesClassifier, ExtraTreesRegressor
+    >>> np.random.seed(0)
+    >>> X = np.random.randn(10000, 4)
+    >>> y = ((X[:, 0]>0) & (X[:, 1]>0)) * np.abs(X[:, 2] * X[:, 3]**2)
+    >>> z = ZeroInflatedRegressor(
+    ... classifier=ExtraTreesClassifier(random_state=0),
+    ... regressor=ExtraTreesRegressor(random_state=0)
+    ... )
+    >>> z.fit(X, y)
+    ZeroInflatedRegressor(classifier=ExtraTreesClassifier(random_state=0),
+                          regressor=ExtraTreesRegressor(random_state=0))
+    >>> z.predict(X)[:5]
+    array([4.91483294, 0.        , 0.        , 0.04941909, 0.        ])
+    """
+
+    def __init__(self, classifier, regressor):
+        """Initialize."""
+        self.classifier = classifier
+        self.regressor = regressor
+
+    def fit(self, X, y):
+        """
+        Fit the model.
+
+        Parameters
+        ----------
+        X : np.ndarray of shape (n_samples, n_features)
+            The training data.
+
+        y : np.ndarray, 1-dimensional
+            The target values.
+
+        Returns
+        -------
+        ZeroInflatedRegressor
+            Fitted regressor.
+        """
+        X, y = check_X_y(X, y)
+        self._check_n_features(X, reset=True)
+
+        self.classifier_ = clone(self.classifier)
+        self.classifier_.fit(X, y != 0)
+
+        non_zero_indices = np.where(self.classifier_.predict(X) == 1)[0]
+
+        if non_zero_indices.size > 0:
+            self.regressor_ = clone(self.regressor)
+            self.regressor_.fit(X[non_zero_indices], y[non_zero_indices])
+        else:
+            self.regressor_ = None
```

Comment:
I think this should either throw a warning or an exception.
I'm leaning towards an exception since I don't think having an unfitted regressor would ever be the behaviour you wanted out of this model, but I'm happy to be convinced that it could be useful and a warning suffices

Reply:
How about this:

```python
        if non_zero_indices.size > 0:
            self.regressor_ = clone(self.regressor)
            self.regressor_.fit(X[non_zero_indices], y[non_zero_indices])
        else:
            self.regressor_ = DummyRegressor(strategy="constant", constant=0)
            self.regressor_.fit(X, y)
```

including some warning that the regressor got overridden because there were no training samples for it? The regressor won't get called anyway for sensible classifiers since the model will output zero all the time.

Reply:
I changed it in the PR.

Reply:
I would prefer an exception too. Warnings are too often ignored, especially when run from the command line and we really want to make the user aware if we never predict anything but a zero.

Reply:
Done.